### ISSUE TITLE
suggestion: simplify archetype definition customisation

### ISSUE THREAD

**Issue by jaredfholgate** [Azure/terraform-provider-alz#38]
With the new module / provider, we are still providing two methods to specify / modify archetypes.

Either by supplying your own library or by supplying variables to add / remove from built in / custom archetypes.

I think we could simplify the provider / module by just supporting a single method of customisation.

I would propose enabling overriding / adding to the built in archetypes. If someone supplies a custom library that includes archetypes and they have the `use_alz_lib` set to `true`, then we should take the archetype definitions in their folder and either override the built in archetype definitions if the name matches or add a new archetype if it does not already exist. We just take the first or last if they provide multiple folders with the same ones (or throw an error if we like).

By doing this, we offer full flexibility while still allowing them to re-use our built in archetypes.

We can then remove the `..._to_add` and `..._to_remove` variables as they are not required for customisation anymore.

I think having one approach is more intuitive and using the files is easier to see / review than the variables. I also think it will simplify the code and documentation for contributors.

This is one for further discussion and as usual it is highly likely I have missed the reason for providing two methods of customisation.

**Comment by matt-FFFFFF** [comment]
The idea behind the `..._to_add` was so that the caller could customise the archetypes without having to maintain a fork.

E.g. they wanted `root` but with the addition of x, y, & z.

**Comment by jaredfholgate** [comment]
Understood, I totally get the point. I think my thoughts are moving towards this:

Using the lib folder is a great way to see what you are customising very easily and works really well with source control. Whereas using vars can be less obvious, especially if combined with a custom lib. This is because customers can supply vars in multiple different ways and even end up building their own custom config file structure to manage it. 

I was thinking that perhaps there could be a concept of an override archetype where you can use the json file to inherit from another archetype (built in or custom) and add / remove things using that file structure instead of using the vars. I appreciate that the removing aspect needs some thought as to how it would be represented in the json.

In summary, having a single opinionated way to customise archetypes that is easy to see and understand. Providing flexibility is great, but having two ways to achieve the same result is where I think we could end up accidently causing confusion.

**Comment by matt-FFFFFF** [comment]
Suggestion accepted! It's on the roadmap. Thanks mate!

**Comment by matt-FFFFFF** [comment]
closed as released in v0.10.0
